### PR TITLE
Baremetal root password and controlplane neutron configs

### PR DESCRIPTION
**Pull request by Jaganathancse**
Adding below changes
1. Baremetal root password secret creation based on base64 password
2. Controlplane Neutron custom service configs changes
3.  Custom Nova openstack dataplane service with configmaps for baremetal compute nodes.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -700,6 +702,7 @@ edpm_deploy_generate_keys:
 	$(eval $(call vars,$@,dataplane))
 	devsetup/scripts/gen-ansibleee-ssh-key.sh
 	bash scripts/gen-edpm-nova-migration-ssh-key.sh
+        bash scripts/gen-edpm-bm-root-password-secret.sh
```
Comment (rabi) [inline Makefile:705]: Run this script only if $BM_ROOT_PASSWORD has been set.

Comment (Jaganathancse) [reply Makefile:705]: > Run this script only if $BM_ROOT_PASSWORD has been set.

ack

Comment (Jaganathancse) [reply Makefile:707]: > > Run this script only if $BM_ROOT_PASSWORD has been set.
> 
> ack

done

**Comment by abays** [comment]
> Build failed (check pipeline). Post `recheck` (without leading slash) to rerun all jobs. Make sure the failure cause has been resolved before you rerun jobs.
> 
> https://review.rdoproject.org/zuul/buildset/71617376c5e0462ab64ba2624d698daa
> 
> ✔️ [openstack-k8s-operators-content-provider ](https://review.rdoproject.org/zuul/build/8bd55bc88ec445779cf0efff0f4c1e9e) SUCCESS in 51m 23s ❌ [install-yamls-crc-podified-edpm-baremetal ](https://review.rdoproject.org/zuul/build/6f9deec1b021499293e11ed05bc9edc8) **FAILURE** in 34m 08s ❌ [podified-multinode-edpm-deployment-crc ](https://review.rdoproject.org/zuul/build/b2502b9371564da8b69acbfcb9ae5dd0) **FAILURE** in 30m 48s ❌ [cifmw-data-plane-adoption-osp-17-to-extracted-crc ](https://review.rdoproject.org/zuul/build/0d962750aadf48c4acba6a934fa9fdc5) **RETRY_LIMIT** in 35m 46s

Looks like this PR is feeding bad config into the system:

```
Monday 19 February 2024  05:40:54 -0500 (0:00:04.996)       0:15:10.496 ******* 
fatal: [localhost]: FAILED! => {"changed": false, "details": "error: trouble configuring builtin PatchTransformer with config: `\npatch: |-\n  - op: add\n    path: /spec/neutron/template/customServiceConfig\n    value: \"\"\"\"\ntarget:\n  kind: OpenStackControlPlane\n`: unable to parse SM or JSON patch from [- op: add\n  path: /spec/neutron/template/customServiceConfig\n  value: \"\"\"\"]\n", "error": "trouble configuring builtin PatchTransformer with config: `", "kustomization": {"apiVersion": "kustomize.config.k8s.io/v1beta1", "kind": "Kustomization", "namespace": "openstack", "patches": [{"patch": "- op: replace\n  path: /spec/secret\n  value: osp-secret\n- op: replace\n  path: /spec/storageClass\n  value: \"local-storage\"", "target": {"kind": "OpenStackControlPlane"}}, {"patch": "apiVersion: core.openstack.org/v1beta1\nkind: OpenStackControlPlane\nmetadata:\n  name: unused\nspec:\n  ovn:\n    template:\n      ovnController:\n        nicMappings:\n          datacentre: ospbr", "target": {"kind": "OpenStackControlPlane"}}, {"patch": "- op: add\n  path: /spec/neutron/template/customServiceConfig\n  value: \"\"\"\"", "target": {"kind": "OpenStackControlPlane"}}], "resources": ["cifmw-kustomization-result.yaml"]}, "kustomization_path": "/home/zuul/ci-framework-data/artifacts/manifests/openstack/openstack/cr/cifmw-kustomize-workspace/kustomization.yaml"}
```

**Comment by Jaganathancse** [comment]
recheck

File: Makefile [hunk 2]
```diff
@@ -407,9 +410,9 @@ DATAPLANE_KUTTL_CONF                             ?= ${OPERATOR_BASE_DIR}/datapla
 DATAPLANE_KUTTL_DIR                              ?= ${OPERATOR_BASE_DIR}/dataplane-operator/tests/kuttl/tests
 DATAPLANE_KUTTL_NAMESPACE                        ?= dataplane-kuttl-tests
 BM_CTLPLANE_INTERFACE                            ?= enp1s0
-BM_ROOT_PASSWORD_SECRET                          ?=
+BM_ROOT_PASSWORD                                 ?=
 GENERATE_SSH_KEYS				 ?= true
-DATAPLANE_EXTRA_NOVA_CONFIG_FILE                 ?= /dev/null
+DATAPLANE_EXTRA_NOVA_CONFIG_FILE                 ?=
```
Comment (gibizer) [inline Makefile:415]: while it is a problem if we create an empty extra config by default?

**Comment by Jaganathancse** [comment]
recheck

File: Makefile [hunk 3]
```diff
@@ -791,7 +805,9 @@ edpm_deploy_prep: edpm_deploy_cleanup ## prepares the CR to install the data pla
 	cp ${DATAPLANE_SERVICE_NOVA_CR} ${DEPLOY_DIR}
 	cp ${DATAPLANE_NODESET_CR} ${DEPLOY_DIR}
 	cp ${DATAPLANE_DEPLOYMENT_CR} ${DEPLOY_DIR}
+ifneq ($(DATAPLANE_EXTRA_NOVA_CONFIG_FILE),)
 	cp ${DATAPLANE_EXTRA_NOVA_CONFIG_FILE} ${EDPM_EXTRA_NOVA_CONFIG_FILE}
+endif
```
Comment (gibizer) [inline Makefile:810]: I think this is an unnecessary change if you keep the default for the env variable

File: Makefile [hunk 4]
```diff
@@ -826,27 +842,37 @@ edpm_deploy_baremetal_prep: export EDPM_NTP_SERVER=${DATAPLANE_NTP_SERVER}
...
+edpm_deploy_baremetal_prep: export EDPM_EXTRA_NOVA_CONFIG_FILE=${DEPLOY_DIR}/25-nova-extra.conf
+edpm_deploy_baremetal_prep: export EDPM_SERVER_ROLE=compute
 edpm_deploy_baremetal_prep: edpm_deploy_cleanup ## prepares the CR to install the data plane
 	$(eval $(call vars,$@,dataplane))
 	mkdir -p ${OPERATOR_BASE_DIR} ${OPERATOR_DIR} ${DEPLOY_DIR}
 	bash scripts/clone-operator-repo.sh
 	cp devsetup/edpm/services/* ${OPERATOR_BASE_DIR}/${OPERATOR_NAME}-operator/config/services
 	cp ${DATAPLANE_NODESET_BAREMETAL_CR} ${DEPLOY_DIR}
 	cp ${DATAPLANE_DEPLOYMENT_BAREMETAL_CR} ${DEPLOY_DIR}
+ifneq ($(DATAPLANE_EXTRA_NOVA_CONFIG_FILE),)
```
Comment (gibizer) [inline Makefile:859]: ditto

File: scripts/gen-edpm-nova-custom-kustomize.sh
Comment (gibizer) [inline scripts/gen-edpm-nova-custom-kustomize.sh:1]: I guess you copied it from scripts/gen-edpm-kustomize.sh,  if so the you should probably remove it from there and call this at the necessary places. Bottom line, it is OK to move the logic, to make it easier to reuse, but it is not OK to duplicate the logic.

File: scripts/gen-edpm-nova-custom-kustomize.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+set -ex
+
+
+if [ "${EDPM_SERVER_ROLE}" == "compute" ]; then
```
Comment (gibizer) [inline scripts/gen-edpm-nova-custom-kustomize.sh:19]: please either indent the body of the if, or rephrase it as as guard condition with a return instead

File: Makefile [hunk 6]
```diff
@@ -826,27 +842,37 @@ edpm_deploy_baremetal_prep: export EDPM_NTP_SERVER=${DATAPLANE_NTP_SERVER}
...
 	make edpm_deploy_generate_keys
 endif
 
 .PHONY: edpm_deploy_baremetal
+edpm_deploy_baremetal: export EDPM_SERVER_ROLE=compute
+edpm_deploy_baremetal: export EDPM_EXTRA_NOVA_CONFIG_FILE=${DEPLOY_DIR}/25-nova-extra.conf
 edpm_deploy_baremetal: input edpm_deploy_baremetal_prep ## installs the dataplane instance using kustomize. Runs prep step in advance. Set DATAPLANE_REPO and DATAPLANE_BRANCH to deploy from a custom repo.
 	$(eval $(call vars,$@,dataplane))
 	oc apply -f ${OPERATOR_BASE_DIR}/${OPERATOR_NAME}-operator/config/services
+ifneq ($(DATAPLANE_EXTRA_NOVA_CONFIG_FILE),)
```
Comment (gibizer) [inline Makefile:873]: you dont need the conditional if you keep the defaulting of the env var

**Review by gibizer — CHANGES_REQUESTED** [review]

File: Makefile (hunk 6 above)
Comment (Jaganathancse) [reply Makefile:872]: @gibizer Ack

File: Makefile (hunk 4 above)
Comment (Jaganathancse) [reply Makefile:859]: Ack

File: Makefile (hunk 3 above)
Comment (Jaganathancse) [reply Makefile:810]: ack. followed existing code format. will update it.

File: Makefile (hunk 2 above)
Comment (Jaganathancse) [reply Makefile:415]: actually nova extra configmap is not creating when passing file. so only  creating configmap when file name is provided

File: scripts/gen-edpm-nova-custom-kustomize.sh
Comment (Jaganathancse) [reply scripts/gen-edpm-nova-custom-kustomize.sh:1]: Ack. will update it.

File: scripts/gen-edpm-nova-custom-kustomize.sh (hunk 5 above)
Comment (Jaganathancse) [reply scripts/gen-edpm-nova-custom-kustomize.sh]: ok

File: Makefile (hunk 6 above)
Comment (Jaganathancse) [reply Makefile]: > you dont need the conditional if you keep the defaulting of the env var

done

File: scripts/gen-edpm-nova-custom-kustomize.sh
Comment (Jaganathancse) [reply scripts/gen-edpm-nova-custom-kustomize.sh:1]: > I guess you copied it from scripts/gen-edpm-kustomize.sh, if so the you should probably remove it from there and call this at the necessary places. Bottom line, it is OK to move the logic, to make it easier to reuse, but it is not OK to duplicate the logic.

done but not supported to rename nova dataplane service as nova-custom, so updated only nova dataplane service for supporting nova-extra-conf service configmap.

File: Makefile (hunk 4 above)
Comment (Jaganathancse) [reply Makefile]: > ditto

done

File: Makefile (hunk 2 above)
Comment (Jaganathancse) [reply Makefile]: > while it is a problem if we create an empty extra config by default?

changed to original.

**Comment by Jaganathancse** [comment]
@gibizer Please review this PR.

File: scripts/gen-edpm-baremetal-kustomize.sh [hunk 7]
```diff
@@ -123,6 +123,47 @@ EOF
...
+# adding more services or splitting existing services.
+# The kustozmization would be something like:
+#     - op: replace
+#      path: /spec/services/11
+#      value: nova-custom
+#
+# So we do a replace by value with yq (assuming golang implementation of yq)
+yq -i '(.spec.services[] | select(. == "nova")) |= "nova-custom"' *openstackdataplanenodeset*.yaml
+fi
+
```
Comment (gibizer) [inline scripts/gen-edpm-baremetal-kustomize.sh:166]: afar as I see this is the same as https://github.com/openstack-k8s-operators/install_yamls/blob/main/scripts/gen-edpm-kustomize.sh#L197-L236. Could you please put this logic to a common place instead of duplicating it?

**Review by gibizer — CHANGES_REQUESTED** [review]

File: scripts/gen-edpm-baremetal-kustomize.sh (hunk 7 above)
Comment (Jaganathancse) [reply scripts/gen-edpm-baremetal-kustomize.sh:166]: ack. will update it. Thanks.

**Review by gibizer — APPROVED** [review]
nova related parts looks good to me

**Comment by gibizer** [comment]
(I've removed the approved label to let others review the neutron and baremetal parts)

File: scripts/gen-edpm-bmset-password-secret.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+# under the License.
+set -ex
+
+function create_bmset_password_secret {
+cat <<EOF | oc create -f -
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: baremetalset-password-secret
```
Comment (rabi) [inline scripts/gen-edpm-bmset-password-secret.sh:24]: This name could have been something that tells that's it's root password.

**Review by rabi — APPROVED** [review]
/lgtm

**Comment by Jaganathancse** [comment]
recheck

File: scripts/gen-service-kustomize.sh [hunk 9]
```diff
@@ -65,6 +65,14 @@ patches:
       path: /spec/storageClass
       value: ${STORAGE_CLASS}
 EOF
+if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}" != "" ]]; then
+    cat <<EOF >>kustomization.yaml
+    - op: add
+      path: /spec/neutron/template/customServiceConfig
+      value: |-
+$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
```
Comment (rdiazcam) [inline scripts/gen-service-kustomize.sh:73]: I think this forces to keep the indentation of kustomization.yaml in the patch file provided. I believe this is a better option (the user should be agnostic about the format of the generated kustomization.yaml):
~~~
diff --git a/scripts/gen-service-kustomize.sh b/scripts/gen-service-kustomize.sh
index 5616334..3a50f7d 100644
--- a/scripts/gen-service-kustomize.sh
+++ b/scripts/gen-service-kustomize.sh
@@ -70,7 +70,7 @@ if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}"
     - op: add
       path: /spec/neutron/template/customServiceConfig
       value: |-
-$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
+      $(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
 EOF
 fi
~~~

Comment (rdiazcam) [reply scripts/gen-service-kustomize.sh:73]: The above snippet only indent the first line of the file patch content. Something like this could valid:
~~~
diff --git a/scripts/gen-service-kustomize.sh b/scripts/gen-service-kustomize.sh
index 5616334..3d6da86 100644
--- a/scripts/gen-service-kustomize.sh
+++ b/scripts/gen-service-kustomize.sh
@@ -70,8 +70,11 @@ if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}"
     - op: add
       path: /spec/neutron/template/customServiceConfig
       value: |-
-$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
 EOF
+while IFS= read -r line
+do
+    echo "      $line" >>kustomization.yaml
+done < ${OPENSTACK_NEUTRON_CUSTOM_CONF}
 fi
 
 IFS=',' read -ra IMAGES <<< "$IMAGE"
~~~
Probably there are more elegant ways of doing this.

**Comment by Jaganathancse** [comment]
recheck

File: scripts/gen-service-kustomize.sh (hunk 9 above)
Comment (Jaganathancse) [reply scripts/gen-service-kustomize.sh:73]: > I think this forces to keep the indentation of kustomization.yaml in the patch file provided. I believe this is a better option (the user should be agnostic about the format of the generated kustomization.yaml):
> 
> ```
> diff --git a/scripts/gen-service-kustomize.sh b/scripts/gen-service-kustomize.sh
> index 5616334..3a50f7d 100644
> --- a/scripts/gen-service-kustomize.sh
> +++ b/scripts/gen-service-kustomize.sh
> @@ -70,7 +70,7 @@ if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}"
>      - op: add
>        path: /spec/neutron/template/customServiceConfig
>        value: |-
> -$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
> +      $(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
>  EOF
>  fi
> ```

Tried this also but no changes in kustomization.yaml.

Comment (Jaganathancse) [reply scripts/gen-service-kustomize.sh:73]: > The above snippet only indent the first line of the file patch content. Something like this could valid:
> 
> ```
> diff --git a/scripts/gen-service-kustomize.sh b/scripts/gen-service-kustomize.sh
> index 5616334..3d6da86 100644
> --- a/scripts/gen-service-kustomize.sh
> +++ b/scripts/gen-service-kustomize.sh
> @@ -70,8 +70,11 @@ if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}"
>      - op: add
>        path: /spec/neutron/template/customServiceConfig
>        value: |-
> -$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
>  EOF
> +while IFS= read -r line
> +do
> +    echo "      $line" >>kustomization.yaml
> +done < ${OPENSTACK_NEUTRON_CUSTOM_CONF}
>  fi
>  
>  IFS=',' read -ra IMAGES <<< "$IMAGE"
> ```
> 
> Probably there are more elegant ways of doing this.

I will place another PR to fix this alignment issue. will merge this PR, getting merge conflicts frequently.

Comment (Jaganathancse) [reply scripts/gen-service-kustomize.sh:73]: > > The above snippet only indent the first line of the file patch content. Something like this could valid:
> > ```
> > diff --git a/scripts/gen-service-kustomize.sh b/scripts/gen-service-kustomize.sh
> > index 5616334..3d6da86 100644
> > --- a/scripts/gen-service-kustomize.sh
> > +++ b/scripts/gen-service-kustomize.sh
> > @@ -70,8 +70,11 @@ if [[ "${KIND}" == "OpenStackControlPlane" && "${OPENSTACK_NEUTRON_CUSTOM_CONF}"
> >      - op: add
> >        path: /spec/neutron/template/customServiceConfig
> >        value: |-
> > -$(cat ${OPENSTACK_NEUTRON_CUSTOM_CONF})
> >  EOF
> > +while IFS= read -r line
> > +do
> > +    echo "      $line" >>kustomization.yaml
> > +done < ${OPENSTACK_NEUTRON_CUSTOM_CONF}
> >  fi
> >  
> >  IFS=',' read -ra IMAGES <<< "$IMAGE"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Probably there are more elegant ways of doing this.
> 
> I will place another PR to fix this alignment issue. will merge this PR, getting merge conflicts frequently.

but tested this PR and alignment is not a problem when using with value: |-

**Review by rabi — APPROVED** [review]
/lgtm
